### PR TITLE
Fix: add custom attribute to the filtered out tabs to keep track

### LINKED ISSUE #999: Search field cleared when I try to delete a tab
To reproduce:

* Search for something (e.g. "linux") that only a subset of your tab titles will have
* Attempt to remove one of those tabs while search filter is in place

Expected:

* Tab will be removed, search filter stays in place

Actual:

* Search filter gets cleared, tab not removed

### LINKED ISSUE #1040: Tabs thumbnail view forced from condensed view if filtering is applied and a new tab is opened
**[Affected versions]:**
- Firefox 52.0.2 and above
- Tab Center 1.31.0 (release version)
- Tab Center custom repo build 

**[Affected Platforms]:**
- All Windows
- All Mac
- All Linux

**[Prerequisites]:**
- Tab Center was installed from Test Pilot or from custom repository build.

**[Steps to reproduce]:**
1. Open the browser with the profile from prerequisites.
2. Navigate to a website with multiple links (eg: wikipedia.org) 
3. Open multiple links in different tabs until the sidebar switches to condensed view.
4. Click the Search input field from Tab Center and enter a string contained or not in the tabs titles.
5. Click the "New Tab" button and observe the tabs behavior.

**[Expected result]:**
A new tab page is opened and focused in the same condensed view mode.

**[Actual result]:**
A new tab page is opened but it's not focused and it forces all the tabs display in thumbnail view mode. 

**[Notes]:**
- The view does not fall back to condensed view even if mouse is moved out of the sidebar.
- This issue also occurs when opening the "about:preferences" or "about:addons" pages from hamburger menu but it does not occur if you open another link from a page in a new tab.
- The workaround for the issue is to close the newly opened page (tab).
- The issue is not reproducible if tabs are in reverse.
- See attached screen recording of the issue: https://goo.gl/ZW2UQF.

### LINKED ISSUE #1043: Search bug - Losing all open tabs
So, I had 119 tabs open, and searched for a particular keyword: "XYZ".
The search list gave me 6 tabs, and I ended up closing all 6 of them. Search remains blank, and no other tabs are shown. I press `escape`, try opening a new tab, but TabCenter is still empty, and all tabs seem lost.
I quit firefox, and started it again, and then hit restore session, only my pinned tabs are being restored. Any way to get back my tabs?

### LINKED ISSUE #1044: Dragged tabs disappear and sidebar wrongly displays the remaining ones if you rearrange tabs while filtering
[Notes]:
- Due to multiple issues observed with reordering while filtering, maybe we should consider disabling this. In my opinion this only complicates the functionality and the benefit is very small.

**[Affected versions]:**
- Firefox 52.0.2 and above
- Tab Center 1.32.0 (release version)
- Tab Center custom repo build 

**[Affected Platforms]:**
- All Windows
- All Mac
- All Linux

**[Prerequisites]:**
- Tab Center was already installed on a new profile from https://testpilot.firefox.com website.
- From a page that contains embedded links (eg: https://en.wikipedia.org), several links were already opened multiple times (in order to have duplicates).

**[Steps to reproduce]:**
1. Open the browser with the profile from prerequisites.
2. Click on Tab Center search field and filter the tabs with a string that will display multiple different tabs.
3. Rearrange at least one tab while filtering is active.
4. Delete the filtering string and enter a string that will only show a specific duplicated tab.
5. Rearrange one of the tabs and observe the behavior.
6. Delete the filtering string and observe the sidebar tabs behavior.

**[Expected result]:**
Step 5: The filtered tabs cannot be rearranged.
Step 6: The sidebar tabs order remains the same. 

**[Actual result]:**
Step 5: The selected tab wrongly disappears when moved.
Step 6:  The tabs opened in the sidebar are wrongly displayed and some of them are overlapping.

**[Notes]:**
- If all the tabs are closed afterwards, you remain with an empty sidebar.
- If you drag down 2 of the tabs from the broken list, the tabs are correctly rearranging.
- When this issue occurs, usually afterwards you are unable to rearrange tabs in filtering mode again.
- See attached screen recording of the issue: https://goo.gl/sRXki6.

### LINKED ISSUE #1045: Dragged tab is wrongly positioned if you previously rearrange filtered tabs
[Notes]:
- If the tabs are not intertwined as described in prerequisites, the dragged tab from Step 4 will not go past the previously filtered tabs.
- If you drag one of the tabs from the broken list, the tabs are correctly rearranged.
- The issue is also reproducible with Tabs in reverse.
- The issue is not reproducible if tabs are in small view.

[Affected versions]:
- Firefox 52.0.2 and above
- Tab Center 1.32.0 (release version)
- Tab Center custom repo build 

[Affected Platforms]:
- All Windows
- All Mac
- All Linux

[Prerequisites]:
- Tab Center was already installed on a new profile from https://testpilot.firefox.com website.
- From a page that contains embedded links (eg: https://en.wikipedia.org), several (eg. 4) links were already opened.
- Several other (different) tabs (eg. 3) are between the ones above.

[Steps to reproduce]:
1. Open the browser with the profile from prerequisites.
2. Pin 2 Wikipedia tabs and write "wiki" in the Search box.
3. Rearrange the pinned tabs and one of the other Wikipedia tabs.
4. Press "ESC" and drag one of the non-wikipedia tabs down to the bottom of the Sidebar.
5. Release the mouse button and observe the Sidebar. 

[Expected result]:
- The dragged tab is placed in the last position of the Sidebar.

[Actual result]:
- The dragged tab is placed at the bottom of the Sidebar, leaving a lot of empty space between the second to last tab and the dragged one.

[Additional notes]:
- See attached screen recording of the issue:
dragged tab appears in the wrong place after you rearrange tabs while filtering

### LINKED ISSUE #1046: "Close Tabs Below/Above" context menu options close all filtered out tabs
[Notes]:
- We should consider disabling the two options if tabs are filtered. Or make the options affect only the tabs which are in view and not the ones which were filtered out.
- The issue is also reproducible if tabs are in reverse order.

[Affected versions]:
- Firefox 52.0.2 and above
- Tab Center 1.32.0 (release version)
- Tab Center custom repo build 

[Affected Platforms]:
- All Windows
- All Mac
- All Linux

[Prerequisites]:
- Tab Center was already installed on a new profile from https://testpilot.firefox.com website.
- From a page that contains embedded links (eg: https://en.wikipedia.org), several (eg. 4) links were already opened.
- Several other (different) tabs (eg. 3) are between the ones above.

[Steps to reproduce]:
1. Open the browser with the profile from prerequisites.
2. Write "wiki" in the Search box.
3. Right click the second to last tab and click the "Close Tabs Below" option.
4. Confirm the close tabs dialog if needed and observe the Sidebar.

[Expected result]:
- Search filter is cleared and the Tabs which were below the selected one are no longer displayed.

[Actual result]:
- The tabs which were filtered out are all closed.

[Additional notes]:
- See attached screen recording of the issue:
close tabs below or above closes all filtered out tabs

### PR DESCRIPTION
**Pull request by ericawright**
Fix many search bugs. Also, show previously filtered tabs if the last visible tab closes.

Fixes #1043.
Fixes #1045.
Fixes #1044.
Fixes #1046.
Fixes #999.
Fixes #1040.

#### Todo 
- ~~double-check that it is still compatible with Tab Groups~~ (yes, minus a little bug)

### CODE REVIEW

File: verticaltabs.js [hunk 1]
```diff
@@ -288,7 +304,7 @@ VerticalTabs.prototype = {
                       window.VerticalTabs.numPinnedtabs();
         break;
       case this.closingTabsEnum.OTHER:
-        tabsToClose = this.visibleTabs.length - 1 - window.VerticalTabs.numPinnedtabs();
+        tabsToClose = this.window.gBrowser.visibleTabs.length - 1 - window.VerticalTabs.numPinnedtabs();
```
Comment (bwinton) [inline verticaltabs.js]: Is `this.window` different than `window`?  And if it isn't, can we just use `window` to match the rest of this line?

File: verticaltabs.js [hunk 2]
```diff
@@ -1174,6 +1191,7 @@ VerticalTabs.prototype = {
     } else {
       hidden_tab.setAttribute('hidden', 'true');
     }
+    this.window.gBrowser._visibleTabs = null;
```
Comment (bwinton) [inline verticaltabs.js]: A comment on why we're doing this would be good, I feel.

**Comment by bwinton** [comment]
Testing revealed the following oddities:

**STR:**
1. Install Tab Center, then Install Tab Groups.
2. Open some tabs in each group, and pick a group.
3. Filter until you have one tab in the group, and close it.

More similar oddities with having a single pinned tab, or moving the last tab into a different group.

**Comment by ericawright** [comment]
I don't think it's worth fixing the tab groups bug. It's not all that disturbing and would involve a complicated fix.

**Pull request merged**